Math/NumberTheory/Primes/Factorisation/QuadraticSieve.hs

```diff
@@ -0,0 +1,143 @@
+  let
+    indexedFactorisations = V.toList (findSmoothNumbers sievingIntervalF)
+    solutionBasis = gaussianElimination indexedFactorisations
+    unsignedFactorisations = map (second primeSet) indexedFactorisations
+
+  pure $ map (\sol -> (findFirstSquare n startingPoint sol, findSecondSquare n unsignedFactorisations sol)) solutionBasis
+
+-- Generating sieving interval. This consists of tuples whose first
+-- component is x^2 - n as x runs from the square root of n for a
+-- total of length t. The second component stores the factorisation
+-- modulo 2 as an IntSet.
+generateInterval :: (Integer -> Integer) -> Integer -> Int -> V.Vector (Integer, SignedPrimeIntSet)
+generateInterval f startingPoint dim = V.map (\x -> (x, isNegative x)) vectorOfValues
+  where
+    vectorOfValues = V.generate dim (\i -> f (intToInteger i + startingPoint))
+    isNegative j = SignedPrimeIntSet (j < 0) mempty
+
+-- This algorithm takes the sievingInterval and the factorBase. It divides by
+-- all the primes in the factor base storing the factorisations. The smooth
+-- numbers correspond to tuples whose first component is 1. The second component
+-- is their factorisation.
+smoothSieveM :: MV.MVector s (Integer, SignedPrimeIntSet) -> [Prime Int] -> Integer -> Integer -> ST s ()
+smoothSieveM sievingIntervalM factorBase n startingPoint = do
+  let t = MV.length sievingIntervalM
+  forM_ factorBase $ \prime -> do
+    let modularSquareRoots = sqrtsModPrime n ((fromJust . toPrimeIntegral) prime)
+    forM_ modularSquareRoots $ \modularSquareRoot -> do
+      let startingIndex = integerToInt ((modularSquareRoot - startingPoint) `mod` (intToInteger . unPrime) prime)
+      forM_ [startingIndex, startingIndex + unPrime prime..(t - 1)] $ \entry -> do
+        let change (y, set) = (y `div` (intToInteger . unPrime) prime, prime `insert` set)
+        MV.modify sievingIntervalM change entry
+
+-- This function returns the smooth numbers together with their index. This
+-- is in order to retrieve later the value of x and x^2 - n. The value stored
+-- in the first component of the tuple is a set whose only component is
+-- the index of column before sieving.
+findSmoothNumbers :: V.Vector (Integer, SignedPrimeIntSet) -> V.Vector (S.IntSet, SignedPrimeIntSet)
+findSmoothNumbers = V.imapMaybe selectSmooth
+  where
+    selectSmooth index (residue, factorisation)
+      | residue == 1 = Just (S.singleton index, factorisation)
+      | otherwise    = Nothing
+
+-- This solves the linear equation. It returns a basis for the kernel
+-- of the matrix as a list of IntSet.
+gaussianElimination :: [(S.IntSet, SignedPrimeIntSet)] -> [S.IntSet]
+gaussianElimination [] = []
+gaussianElimination (p@(indices ,pivotFact) : xs) = case nonZero pivotFact of
+  Just pivot -> gaussianElimination (map (\q@(_, fact) -> if pivot `member` fact then add p q else q) xs)
+  Nothing    -> indices : gaussianElimination xs
+  where
+    add (a, u) (b, v) = ((a S.\\ b) <> (b S.\\ a), xor u v)
```

Comment:
Same comment as for `xor` - perhaps manual zipping will be faster if this is called frequently enough?